kc-distance with branch lengths

awohns opened the pull request:
Modified kc_distance to consider branch lengths via the lambda parameter (0 = only topological comparison, 1 = only branch lengths). Renamed to kc_distance_tree as it compares two trees.
Also added kc_distance_ts, which calcualtes kc_distance across a tree sequence.

jeromekelleher commented on python/tests/test_topology.py:
```diff
@@ -48,6 +48,293 @@ def generate_segments(n, sequence_length=100, seed=None):
+                M[tree_index][u + n] = tree.time(tree.parent(u))
+    return ((1 - lambda_param) * np.linalg.norm(m[0] - m[1])) + (lambda_param * np.linalg.norm(M[0] - M[1]))
+
+def kc_distance_ts(ts_1, ts_2, lambda_param = 0):
```
Let's keep it to just trees for the PR @awohns, we can think about the TS case later.

jeromekelleher commented:
What's the status with this PR @awohns? It's slipped under my radar for a bit, let's try and get it merged soon. Looks like the issue that the moment is some linting problems - try running flake8 locally.

Should also rebase to bring this up to date (ask Yan or me for guidance if you're not sure how to do this)

awohns commented:
Ok, I squashed my commits and rebased to tskit/master, hopefully that was all done correctly. The tests all pass and linting is fine now. Let me know if this is alright or if I should make some changes! 
Also, this is only the Python implementation, I haven't gotten around to the C one yet but it's on my todo list.

jeromekelleher commented:
Great, thanks. Can you grab me at some point to go through this with you please?

awohns commented:
Happy to walk you through this anytime!

brianzhang01 commented on python/tests/test_topology.py:
```diff
@@ -49,6 +49,532 @@ def generate_segments(n, sequence_length=100, seed=None):
+def kc_distance_tree(tree1, tree2, topo_v_age=0):
+    """
+    Returns the Kendall-Colijn distance between the specified
+    pair of trees. topology_vs_age determines weight of topology vs
```
Discrepancy between `topo_v_age` and `topology_v_age`

awohns replied:
Good catch

brianzhang01 commented on python/tests/test_topology.py:
```diff
@@ -49,6 +49,532 @@ def generate_segments(n, sequence_length=100, seed=None):
+            if len(tree.children(u)) == 0:
+                M[tree_index][u + n] = tree.branch_length(u)
+    return ((1 - topo_v_age) * np.linalg.norm(m[0] - m[1])) + (
+        topo_v_age * np.linalg.norm(M[0] - M[1]))
```
Are you sure this is equivalent to
```
np.linalg.norm((1 - topo_v_age) * (m[0] - m[1]) + topo_v_age * (M[0] - M[1]))
```
? Because that's what I understood the metric to be. From thinking about it quickly, I don't think the two expressions are the same, because you can't distribute `np.linalg.norm` aka Euclidean distance over a `+` operation.

awohns replied:
Just looked into this, you're right: they're generally not the same thing. They're only the same when ||A+B|| is orthogonal, and that only happens when the topologies between the two trees your comparing is the same. In practice the difference between the distributed version and the correct version is small in absolute terms, but I'll amend it to be correct. Great catch! Thanks Brian

brianzhang01 commented on python/tests/test_topology.py:
```diff
@@ -49,6 +49,532 @@ def generate_segments(n, sequence_length=100, seed=None):
+                        if u < v:
+                            a, b = u, v
+                        else:
+                            a, b = v, u
```
Alternatively,
```
a = min(u, v)
b = max(u, v)
```

awohns replied:
Yep!

brianzhang01 commented on python/tests/test_topology.py:
```diff
@@ -49,6 +49,532 @@ def generate_segments(n, sequence_length=100, seed=None):
+    """
+    Returns the Kendall-Colijn distance between the specified
+    pair of trees. topology_vs_age determines weight of topology vs
+    branch lengths in calculating the distance.
```
Maybe add a paper citation. This is the one I'm looking at: https://academic.oup.com/mbe/article/33/10/2735/2925548

awohns replied:
Good idea

brianzhang01 commented on python/tests/test_topology.py:
```diff
@@ -49,6 +49,530 @@ def generate_segments(n, sequence_length=100, seed=None):
+    """
+    Returns the Kendall-Colijn distance between the specified pair of trees.
+    topo_v_age determines weight of topology vs branch lengths in calculating
+    the distance. See Kendall & Colijn (2016):
```
Might be worth saying that 0 = all topology, 1 = all age

awohns replied:
good point, changed.

jeromekelleher commented:
Great to have an extra pair of eyes on this, thanks @brianzhang01!

jeromekelleher commented:
@awohns, we should pick this up again soon and finish it up. Can you take a look and see what needs doing please?

awohns commented:
Just added a simple version of the python function (without numpy or itertools).
Will now start writing the C version

jeromekelleher commented on python/tests/test_topology.py:
```diff
@@ -97,6 +98,615 @@ def generate_segments(n, sequence_length=100, seed=None):
+        raise ValueError("Trees must have one root")
+    k = samples.shape[0]
+    n = (k * (k - 1)) // 2
+    m = [[1] * (n + k), [1] * (n + k)]
```
Might as well use np.zeros/np.ones here --- it's no closer to C to use this.

jeromekelleher commented on python/tests/test_topology.py:
```diff
@@ -97,6 +98,615 @@ def generate_segments(n, sequence_length=100, seed=None):
+    n = (k * (k - 1)) // 2
+    m = [[1] * (n + k), [1] * (n + k)]
+    M = [[0] * (n + k), [0] * (n + k)]
+    depth_dict = [[0] * tree1.num_nodes, [0] * tree2.num_nodes]
```
Using "_dict" is confusing here. Maybe "node_depth", and "node_time" so we know what they are mapping to?

jeromekelleher replied:
Might also be helpful to change the terminology a bit: maybe "time_distance" and "path_distance", as each of these is measuring distance from the root, right?

jeromekelleher commented on python/tests/test_topology.py:
```diff
@@ -97,6 +98,615 @@ def generate_segments(n, sequence_length=100, seed=None):
+
+
+def kc_distance_tree_simple(tree1, tree2, topo_v_age=0):
+    """
```
No point in duplicating the docstring here. Just say it's a simpler version of the one above, corresponding to the C version.

jeromekelleher commented on python/tests/test_topology.py:
```diff
@@ -97,6 +98,615 @@ def generate_segments(n, sequence_length=100, seed=None):
+    return np.linalg.norm((1 - topo_v_age) * (m[0] - m[1]) + topo_v_age * (M[0] - M[1]))
+
+
+def kc_distance_tree_simple(tree1, tree2, topo_v_age=0):
```
I don't like the topo_v_age parameter. I think it would be better to use the original terminology of calling it alpha or  beta or whatever, and we can explain what this means in the documentation.

jeromekelleher commented on python/tests/test_topology.py:
```diff
@@ -97,6 +98,615 @@ def generate_segments(n, sequence_length=100, seed=None):
+                m[tree_index][pair_index] = depth_dict[tree_index][mrca]
+                M[tree_index][pair_index] = time_dict[tree_index][mrca]
+
+    vT1 = [(depth * (1 - topo_v_age)) + (topo_v_age * time)
```
You're still leaving a lot to do in the C implementation here. This should be a single loop, not using zip or list comprehensions or anything like that. ``for j in range()`` etc.

jeromekelleher commented on python/tests/test_topology.py:
```diff
@@ -97,6 +98,615 @@ def generate_segments(n, sequence_length=100, seed=None):
+    https://academic.oup.com/mbe/article/33/10/2735/2925548
+    """
+    samples = tree1.tree_sequence.samples()
+    for sample1, sample2 in zip(samples, tree2.tree_sequence.samples()):
```
Don't need to bother with error checking here -- this is just duplicating code from above.

jeromekelleher commented on python/tests/test_topology.py:
```diff
@@ -98,11 +98,11 @@ def generate_segments(n, sequence_length=100, seed=None):
 
 
-def kc_distance_tree(tree1, tree2, topo_v_age=0):
+def kc_distance_tree(tree1, tree2, lambda_param=0):
```
``lambda_param`` is awkward. The convention is usually to call it ``lambda_`` to avoid clashes with built-in symbols.

awohns replied:
Got it, thanks for checking this out!

jeromekelleher commented on python/tests/test_topology.py:
```diff
@@ -173,14 +170,19 @@ def kc_distance_tree_simple(tree1, tree2, topo_v_age=0):
+    vT2 = 0 
+    distance_sum = 0
+
+    for i in range(n + k):
```
Nice!

brianzhang01 commented on python/tests/test_topology.py:
```diff
@@ -97,6 +98,618 @@ def generate_segments(n, sequence_length=100, seed=None):
+    https://academic.oup.com/mbe/article/33/10/2735/2925548
+    """
+    samples = tree1.tree_sequence.samples()
+    if not np.array_equal(samples, tree2.tree_sequence.samples()):
```
Does .samples() give sorted samples by default? If not, we may need to sort them. If yes, I'm guessing this is why you didn't write `tree1.samples()`, maybe you could add a short comment that we want to get the samples via the tree_sequence so that things are sorted.

jeromekelleher replied:
``.samples`` returns the samples in sorted order, yes.

brianzhang01 commented on python/tests/test_topology.py:
```diff
@@ -97,6 +98,618 @@ def generate_segments(n, sequence_length=100, seed=None):
+        raise ValueError("Trees must have one root")
+    k = samples.shape[0]
+    n = (k * (k - 1)) // 2
+    m = [np.ones(n + k), np.ones(n + k)]
```
I'm pretty sure it will also work to do
```
m = np.ones(2, n+k)
M = np.zeros(2, n+k)
```
and may be ever slightly faster

brianzhang01 commented on python/tests/test_topology.py:
```diff
@@ -97,6 +98,618 @@ def generate_segments(n, sequence_length=100, seed=None):
+        while len(stack) > 0:
+            u, depth, time = stack.pop()
+            children = tree.children(u)
+            for v in children:
```
huh, this is clever, doing a DFS from the root. i might give this a try as well.

brianzhang01 commented on python/tests/test_topology.py:
```diff
@@ -97,6 +98,618 @@ def generate_segments(n, sequence_length=100, seed=None):
+    n = (k * (k - 1)) // 2
+    m = [np.ones(n + k), np.ones(n + k)]
+    M = [np.zeros(n + k), np.zeros(n + k)]
+    path_distance = [np.zeros(tree1.num_nodes), np.zeros(tree2.num_nodes)]
```
interesting, is it this way because the trees can have different number of nodes due to polytomies?

awohns replied:
Yes, while the samples have to be the same between the two trees, the internal nodes can be different

brianzhang01 commented on python/tests/test_topology.py:
```diff
@@ -97,6 +98,618 @@ def generate_segments(n, sequence_length=100, seed=None):
+                kc_distance_tree_simple(tree1, tree2),
+                kc_distance_tree_simple(tree1, tree2))
+
+    def test_sample_3(self):
```
I use pytest parameterized tests which basically does this without the code duplication. however we're not using pytest. https://docs.pytest.org/en/latest/parametrize.html

jeromekelleher replied:
Pytest looks like all right, but it'd be a big job moving over now. There's not a lot of motivation for taking this work on.

brianzhang01 replied:
Here's another solution I came across which would fit well, and doesn't require any extra packages: https://docs.python.org/3/library/unittest.html#distinguishing-test-iterations-using-subtests.

From https://stackoverflow.com/questions/32899/how-do-you-generate-dynamic-parameterized-unit-tests-in-python

jeromekelleher replied:
Hey, this is great feature I didn't know about, thanks @brianzhang01! I'll definitely be deploying that one.

awohns commented:
> When the number of samples gets large, doing a top-down search from root may be more efficient (as I think you were doing in the Python code). I've played around with both approaches and have settled on a top-down search. Here, for the samples that have a MRCA at root, I think you end up climbing the path to root several times.

Yes I can see how top-down would be more efficient.

> I also used a trick where in my top-down search, instead of calling a function to get the samples underneath a node, I do a postorder traversal using recursion where subnodes of my node each compute their samples underneath. Then you can either concatenate these lists (as there shouldn't be duplicates) or use something like C++ std::merge if you want the sample list in sorted order.

So, at each node, this could find all the pairs of samples where that node is the MRCA. I suppose you just have to keep track of whether the node is from the left or right subtree?

jeromekelleher commented:
Looks good @awohns. Clang has found a couple of C bugs if you look at the CircleCI output, so they should be fixed.

Otherwise, looking good. All that's left to do is hook this up with the Python API so that we can call it from Python now and to write some docs. I'm happy to take this from here, or I can talk you through the remaining steps and  we get them done on Friday. I'd like to release 0.2.3 then, and it'd be nice to have this functionality in for the release.

jeromekelleher commented:
Also, can you [rebase](https://stdpopsim.readthedocs.io/en/latest/development.html#rebasing) this please?

jeromekelleher commented:
> When the number of samples gets large, doing a top-down search from root may be more efficient (as I think you were doing in the Python code). I've played around with both approaches and have settled on a top-down search. Here, for the samples that have a MRCA at root, I think you end up climbing the path to root several times.

This sounds really interested @brianzhang01 --- perhaps you could send us a Python version of your algorithm, (even better) open a PR on the Python implementations here with your ideas? This is certainly not an optimal way of doing KC distance, it would be good to iterate a bit on the implementation and share ideas.

brianzhang01 commented:
I really just used the ideas demonstrated in the [Python version](https://github.com/tskit-dev/tskit/pull/172/files#diff-5941ccebfc6d10311b8d79e8bc986cffR146) and added my own optimization for getting the samples. Sounds like we're going to release this version of the C algorithm first, but I'd be willing to collaborate on a more efficient C version later. (I'm working in C++ with a TreeSequence representation similar to tskit, and using pybind11 for my bindings.)

awohns commented:
I've written some documentation for adding functions to tskit, do we want that in this PR?

jeromekelleher commented:
> I've written some documentation for adding functions to tskit, do we want that in this PR?

Excellent! No, let's open another PR for that.
